src/Data/Constraint/Deriving/CorePluginM.hs

```diff
@@ -588,7 +588,12 @@ replaceTypeOccurrences told tnew = replace
         = mkSpecForAllTys bndrs $ replace t'
         -- split arrow types
       | Just (at, rt) <- splitFunTy_maybe t
-        = mkFunTy (replace at) (replace rt)
+#if __GLASGOW_HASKELL__ >= 810
+        = mkVisFunTy
```

Comment:
Here we need to make sure that we don't mix predicate/normal type applications. Perhaps, here we can use `ft_af :: Type -> AnonArgFlag` to get the first argument for `mkFunTy`. However, be careful with that partial function, as it does not look through synonyms and stuff (i.e. we would need to apply `coreView`, like `splitTyConApp_maybe` does internally).